indexer/metadata/simulation/iceprod_tools.py

```diff
@@ -0,0 +1,403 @@
+                        if module_d["type"] in ("permanent", "site_temp") and module_d[
+                            "movement"
+                        ] in ("output", "both"):
+                            files.append(
+                                {
+                                    "url": module_d["remote"],
+                                    "iters": tray["iterations"],
+                                    "task": task["name"],
+                                }
+                            )
+        return files
+
+    async def _add_task_info_to_job_config(self, job_config: dataclasses.Job) -> None:
+        """Add `"task_info"` dict to `job_config["options"]`."""
+        if "task" not in job_config["options"]:
+            raise TaskNotFound()
+
+        ret = await self.iceprodv2_rc.request(
+            "GET",
+            f"/datasets/{job_config['options']['dataset_id']}/tasks",
+            {
+                "job_index": job_config["options"]["job"],
+                "keys": "name|task_id|job_id|task_index",
+            },
+        )
+
+        # find matching task
+        task = {}
+        for task in ret.values():
+            if task["name"] == job_config["options"]["task"]:
+                job_config["options"]["task_info"] = task
+                return
+
+        raise TaskNotFound()
+
+    @staticmethod
+    async def _match_outfile_and_add_to_job_config(
+        filepath: str, job_config: dataclasses.Job, job_index: Optional[int],
+    ) -> None:
+        """Add `"task"`, `"job"`, & `"iter"` values to `config["options"]`."""
+        if job_index:  # do we already know what job to look at?
+            job_search: List[int] = [job_index]
+        else:  # otherwise, look at each job from dataset
+            job_search = list(range(job_config["options"]["jobs_submitted"]))
+
+        parser = ExpParser()
+        env = {"parameters": job_config["steering"]["parameters"]}
+
+        def get_path_from_url(f_data: _OutFileData) -> str:
+            url = cast(str, parser.parse(f_data["url"], job_config, env))
+            if "//" not in url:
+                path = url
+            else:
+                path = "/" + url.split("//", 1)[1].split("/", 1)[1]
+            logging.info(f"checking path {path}")
+            return path
+
+        # search each possible file/task
+        possible_outfiles = _IceProdV2Querier._get_outfiles(job_config)
+        for f_data in reversed(possible_outfiles):
+            for job in job_search:
+                for i in range(f_data["iters"]):
+                    logging.info(
+                        f"Searching task: {f_data['task']}, job: {job}, iter: {i}"
+                    )
+                    if get_path_from_url(f_data) == filepath:
+                        logging.info(f"Success on job: {job}, iter: {i}")
+                        job_config["options"].update(
+                            {"task": f_data["task"], "job": job, "iter": i}
+                        )
```

Comment:
One note: you're doing an update on the job_config here, which will apply to the main job_config outside this function.  This may be a problem in `_add_task_info_to_job_config` where it checks for `task`, since that will be set from this.

Maybe you want to make a copy of the job config at some point, so these trial values don't get propagated up except on purpose?

Reply:
I'm not seeing this. I'm intending to only update `job_config` when a match is found (`if get_path_from_url(f_data) == filepath:`) and then immediately return.

Reply:
Ah, so, I'd assumed you set those before calling `get_path_from_url()`.  If you're not, then you might not get any match, since the parser won't be able to resolve the `$(job)` in the url.

Reply:
Oh that *is* big. Fixed it

Reply:
It should work now.  Cleanup is a little ugly, but functional (the `job_config` will have bad values if an exception gets raised in the loops, but that's probably OK).

Reply:
That's true, I've written a vast majority of the indexer to bubble up exceptions, so it's in good company